src/v/storage/mvlog/versioned_log.h

```diff
@@ -0,0 +1,142 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "container/fragmented_vector.h"
+#include "model/offset_interval.h"
+#include "model/record.h"
+#include "storage/mvlog/file.h"
+#include "storage/mvlog/segment_identifier.h"
+#include "storage/ntp_config.h"
+#include "utils/mutex.h"
+
+#include <seastar/core/circular_buffer.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/lowres_clock.hh>
+
+namespace storage::experimental::mvlog {
+
+class readable_segment;
+class segment_appender;
+
+struct active_segment {
+    active_segment(active_segment&) = delete;
+    active_segment(active_segment&&) = delete;
+    active_segment& operator=(active_segment&) = delete;
+    active_segment& operator=(active_segment&&) = delete;
+
+    active_segment(
+      std::unique_ptr<file>, model::offset, segment_id, size_t target_size);
+    ~active_segment();
+
+    std::unique_ptr<file> segment_file;
+    std::unique_ptr<segment_appender> appender;
+    std::unique_ptr<readable_segment> readable_seg;
+
+    // Time this segment was created, used to determine the age of the segment
+    // when applying segment.ms.
+    const ss::lowres_clock::time_point construct_time;
+
+    // Target max size of the segment.
+    const size_t target_max_size;
+
+    const segment_id id;
+    model::offset base_offset;
+    model::offset next_offset;
+};
+
+struct readonly_segment {
+    readonly_segment(readonly_segment&) = delete;
+    readonly_segment(readonly_segment&&) = delete;
+    readonly_segment& operator=(readonly_segment&) = delete;
+    readonly_segment& operator=(readonly_segment&&) = delete;
+
+    explicit readonly_segment(std::unique_ptr<active_segment>);
+    ~readonly_segment();
+
+    std::unique_ptr<file> segment_file;
```

Comment:
does `file` holds file descriptor or it can open/close file dynamically?

Reply:
The underlying file descriptor is managed by the io:: subsystem so long term the responsibility will fall there.